src/compiler/checker.ts

```diff
@@ -1093,38 +1093,75 @@ namespace ts {
+                        };
+                    }
+                }
+                else if (lookupTable && exportNode && id !== "default" && hasProperty(target, id) && resolveSymbol(target[id]) !== resolveSymbol(source[id])) {
+                    lookupTable[id].exportsWithDuplicate.push(exportNode);
                 }
             }
         }
 
         function getExportsForModule(moduleSymbol: Symbol): SymbolTable {
-            let result: SymbolTable;
             const visitedSymbols: Symbol[] = [];
-            visit(moduleSymbol);
-            return result || moduleSymbol.exports;
+            return visit(moduleSymbol) || moduleSymbol.exports;
 
             // The ES6 spec permits export * declarations in a module to circularly reference the module itself. For example,
             // module 'a' can 'export * from "b"' and 'b' can 'export * from "a"' without error.
-            function visit(symbol: Symbol) {
+            function visit(symbol: Symbol): SymbolTable {
                 if (symbol && symbol.flags & SymbolFlags.HasExports && !contains(visitedSymbols, symbol)) {
                     visitedSymbols.push(symbol);
-                    if (symbol !== moduleSymbol) {
-                        if (!result) {
-                            result = cloneSymbolTable(moduleSymbol.exports);
-                        }
-                        extendExportSymbols(result, symbol.exports);
-                    }
+                    const symbols: SymbolTable = cloneSymbolTable(symbol.exports);
                     // All export * declarations are collected in an __export symbol by the binder
                     const exportStars = symbol.exports["__export"];
                     if (exportStars) {
+                        const nestedSymbols: SymbolTable = {};
+                        const lookupTable: Map<ExportCollisionTracker> = {};
                         for (const node of exportStars.declarations) {
-                            visit(resolveExternalModuleName(node, (<ExportDeclaration>node).moduleSpecifier));
+                            const resolvedModule = resolveExternalModuleName(node, (node as ExportDeclaration).moduleSpecifier);
+                            const exportedSymbols = visit(resolvedModule);
+                            extendExportSymbols(
+                                nestedSymbols,
+                                exportedSymbols,
+                                lookupTable,
+                                node as ExportDeclaration
+                            );
+                        }
+                        for (const id in lookupTable) {
+                            const { exportsWithDuplicate } = lookupTable[id];
+                            // Its not an error if the file with multiple export *'s with duplicate names exports a member with that name itself
+                            if (id === "export=" || !exportsWithDuplicate.length || id in symbols) {

```

Comment:
Maybe it should be, but there are places where we probably use `in` *in*tentionally